src/main/java/io/spokestack/spokestack/asr/SpokestackCloudClient.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    /**
+     * @return true if the websocket is currently connected, false otherwise
+     */
+    public boolean isConnected() {
+        return this.socket != null;
+    }
+
+    /**
+     * establishes a websocket connection for speech recognition.
+     */
+    public void connect() {
+        if (this.socket != null) {
+            throw new IllegalStateException();
+        }
+
+        Request request = new Request.Builder().url(SOCKET_URL).build();
+        this.socket = this.client.newWebSocket(request, new SocketListener());
+        this.socket.send(this.authMessage);
+    }
+
+    /**
+     * disconnects the websocket.
+     */
+    public void disconnect() {
+        if (this.socket != null) {
+            try {
+                this.socket.close(1000, "goodbye");
+            } finally {
+                this.socket = null;
+                this.buffer.clear();
+            }
+        }
+    }
+
+    /**
+     * transmits an audio frame over the websocket.
+     *
+     * @param frame the audio frame buffer to send
+     */
+    public void sendAudio(ByteBuffer frame) {
+        if (this.socket == null) {
+            throw new IllegalStateException();
+        }
+
+        if (this.buffer.remaining() < frame.capacity()) {
+            flush();
+        }
+
+        frame.rewind();
+        this.buffer.put(frame);
+    }
+
+    /**
+     * sends any buffered data followed by an empty message indicating the end
+     * of the utterance.
+     */
+    public void endAudio() {
+        if (this.socket == null) {
+            throw new IllegalStateException();
+        }
+
+        flush();
+        this.socket.send(ByteString.EMPTY);
+    }
+
+    private void flush() {
+        if (this.buffer.position() > 0) {
+            this.buffer.flip();
+            this.socket.send(ByteString.of(this.buffer));
+            this.buffer.clear();
+        }
+    }
+
+    /**
+     * okhttp socket callback.
+     */
+    private class SocketListener extends WebSocketListener {
+        private final Gson gson = new Gson();
+        private final Listener listener = SpokestackCloudClient.this.listener;
+
+        @Override
+        public void onMessage(@NotNull WebSocket s, @NotNull String message) {
+            SpokestackASRResponse response =
+                  gson.fromJson(message, SpokestackASRResponse.class);
+            if (response.error != null) {
+                String err = String.format("ASR error: %s", response.error);
+                this.listener.onError(new Exception(err));
+                return;
+            }
+            String speech = "";
+            float confidence = 0.0f;
+            if (response.status.equals("ok") && response.isFinal
+                  && response.hypotheses.length > 0) {
+                speech = response.hypotheses[0].transcript;
+                confidence = response.hypotheses[0].confidence;
+            }
+            this.listener.onSpeech(speech, confidence);
```

Comment:
why send a notification here with an empty transcript if this isn't the final response?

Reply:
As you might have noticed, this is mostly a copy of the old Bing recognizer, which did this. I didn't question the status quo too much, but you're right; the empty one should probably go away.

In a future version, I'd like to add a partial transcript event to the listener so that an app could choose to display "real-time" results like Google Assistant does. Before I remembered to only send a transcript marked as final, I was displaying them all in my test app, and it felt like a better experience—it gives the illusion of the ASR being more responsive or listening more actively. Even then, though, no empty events would be sent.

Reply:
burned by past bs

Reply:
I agree on the partials, might make sense as a new event type